src/realm/object-store/audit.mm

```diff
@@ -1081,45 +1093,83 @@ void validate_metadata(std::vector<std::pair<std::string, std::string>>& metadat
 
 {
     util::CheckedLockGuard lock(m_mutex);
-    auto scope = std::make_shared<Scope>(Scope{m_metadata, std::string(activity)});
+    // scope id isn't used for this scope, so it's just an arbitrary value
+    auto scope = std::make_shared<Scope>(Scope{0, m_metadata, std::string(activity)});
     scope->events.push_back(audit_event::Custom{now(), std::string(activity), event_type, data});
     scope->completion = std::move(completion);
     trigger_write(std::move(scope));
 }
 
 void AuditContext::pin_version(VersionID version)
 {
-    for (auto& transaction : m_current_scope->source_transactions) {
-        if (transaction->get_version() == version.version)
-            return;
+    TransactionRef pin;
+    for (auto& scope : m_current_scopes) {
+        bool has_version =
+            std::any_of(scope->source_transactions.begin(), scope->source_transactions.end(), [&](auto& tr) {
+                return tr->get_version() == version.version;
+            });
+        if (!has_version) {
+            if (!pin) {
+                pin = m_source_db->start_read(version);
+            }
+            scope->source_transactions.push_back(pin);
+        }
+    }
+}
+
+uint64_t AuditContext::begin_scope(std::string_view name)
+{
+    util::CheckedLockGuard lock(m_mutex);
+    auto id = ++m_scope_counter;
+    m_logger->trace("Events: Beginning event scope '%1' on '%2' named '%3'", id, m_source_db->get_path(), name);
+    m_current_scopes.push_back(std::make_shared<Scope>(Scope{id, m_metadata, std::string(name)}));
+    return id;
+}
+
+std::vector<std::shared_ptr<AuditContext::Scope>>::iterator AuditContext::find_scope(uint64_t id)
+{
+    return std::find_if(m_current_scopes.begin(), m_current_scopes.end(), [&](auto& scope) {
+        return scope->id == id;
+    });
+}
+
+void AuditContext::end_scope(uint64_t id, util::UniqueFunction<void(std::exception_ptr)>&& completion)
+{
+    util::CheckedLockGuard lock(m_mutex);
+    auto it = find_scope(id);
+    if (it == m_current_scopes.end()) {
+        throw std::logic_error(util::format(
+            "Cannot end event scope: scope '%1' not in progress. Scope may have already been ended?", id));
     }
-    m_current_scope->source_transactions.push_back(m_source_db->start_read(version));
+    m_logger->trace("Events: Comitting event scope '%1' on '%2' with %3 events", id, m_source_db->get_path(),
+                    (*it)->events.size());
+    (*it)->completion = std::move(completion);
+    trigger_write(std::move(*it));
+    m_current_scopes.erase(it);
```

Comment:
list is the right option here since you only do the linear search, add to the end and here and removal by iterator

Reply:
std::list is almost never the correct option. A linked list's poor memory locality means that a linear find and remove on a std::vector is typically much faster than on a std::list even with std::list::remove() being O(1). The combined operation is O(N) either way.